lib/matplotlib/tests/test_figure.py

```diff
@@ -687,6 +687,85 @@ def test_removed_axis():
     axs[0].remove()
     fig.canvas.draw()
 
+def test_figure_clear():
+    # we test the following figure clearing scenarios:
+    fig = plt.figure()
+    
+    # a) an empty figure
+    fig.clear()
+    assert fig.axes == []
+    
+    # b) a figure with a single unnested axes
+    ax = fig.add_subplot(111)
+    fig.clear()
+    assert fig.axes == []
+    
+    # c) a figure multiple unnested axes
+    axes = [fig.add_subplot(2,1,i+1) for i in range(2)]
+    fig.clear()
+    assert fig.axes == []
+    
+    # d) a figure with a subfigure
+    gs = fig.add_gridspec(ncols=2,nrows=1)
+    subfig = fig.add_subfigure(gs[0])
+    subaxes = subfig.add_subplot(111)
+    fig.clear()
+    assert subfig not in fig.subfigs
+    assert fig.axes == []
+    
+    # e) a figure with a subfigure and a subplot
+    subfig = fig.add_subfigure(gs[0])
+    subaxes = subfig.add_subplot(111)
+    mainaxes = fig.add_subplot(gs[1])
+    
+    # e.1) removing just the axis leaves the subplot
+    mainaxes.remove()
+    assert [subaxes] == fig.axes
+    
+    # e.2) removing just the subaxis leaves the subplot
+    # and subfigure
+    mainaxes = fig.add_subplot(gs[1])
+    subaxes.remove()
+    assert [mainaxes] == fig.axes
+    assert subfig in fig.subfigs
+    
+    # e.3) clearing the subfigure leaves the subplot
+    subaxes = subfig.add_subplot(111)
+    assert all(ax in fig.axes for ax in [mainaxes,subaxes])
+    subfig.clear()
+    assert subfig in fig.subfigs
+    assert subaxes not in subfig.axes
+    assert subaxes not in fig.axes
+    assert mainaxes in fig.axes
+    
+    #e.4) clearing the whole thing
+    subaxes = subfig.add_subplot(111)
+    fig.clear()
+    assert fig.axes == []
+    assert fig.subfigs == []
+
+    #f) multiple subfigures
+    subfigs = [fig.add_subfigure(gs[i]) for i in [0,1]]
+    subaxes = [sfig.add_subplot(111) for sfig in subfigs]
+    assert all(ax in fig.axes for ax in subaxes)
+    assert all(sfig in fig.subfigs for sfig in subfigs)
+    
+    #f.1) clearing only one subfigure
+    subfigs[0].clear()
+    assert subaxes[0] not in fig.axes
+    assert subfigs[0] not in fig.subfigs
```

Comment:
Does clearing a sub fig remove it from the figure?  I'm not following why that is desirable?  The parent should remove it if we want it removed...

Reply:
Clearing  a subfigure should not remove it from the parent figure